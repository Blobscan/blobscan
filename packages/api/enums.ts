export { BlobStorage, Category, Rollup } from "@blobscan/db/prisma/enums";
