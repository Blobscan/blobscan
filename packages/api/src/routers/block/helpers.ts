import type { BlobStorageManager } from "@blobscan/blob-storage-manager";
import type { BlobscanPrismaClient, Prisma } from "@blobscan/db";
import { z } from "@blobscan/zod";

import type {
  ExpandedBlob,
  ExpandedTransaction,
  Expands,
} from "../../middlewares/withExpands";
import type { Filters } from "../../middlewares/withFilters";
import type { Prettify } from "../../types";
import { isFullyDefined } from "../../utils";
import { retrieveBlobData } from "../../utils";
import {
  deriveTransactionFields,
  normalizePrismaBlobFields,
  normalizePrismaTransactionFields,
} from "../../utils/transformers";
import {
  baseBlobSchema,
  baseBlockSchema,
  baseTransactionSchema,
} from "../../zod-schemas";

const prismaBlockSelect = {
  hash: true,
  number: true,
  timestamp: true,
  slot: true,
  blobGasUsed: true,
  blobAsCalldataGasUsed: true,
  blobGasPrice: true,
  excessBlobGas: true,
} satisfies Prisma.BlockSelect;

export type PrismaBlock = Prisma.BlockGetPayload<{
  select: typeof prismaBlockSelect;
}>;

export type CompletePrismaBlock = Prettify<
  PrismaBlock & {
    transactions: Prettify<
      {
        hash: string;
      } & Partial<ExpandedTransaction> & {
          blobs: Prettify<
            { blobHash: string } & {
              blob?: ExpandedBlob;
            }
          >[];
        }
    >[];
  }
>;

export function createBlockSelect(expands: Expands, filters?: Filters) {
  const blobExpand = expands.blob ? { blob: expands.blob } : {};
  const transactionExpand = expands.transaction?.select ?? {};

  return {
    ...prismaBlockSelect,
    transactions: {
      select: {
        hash: true,
        ...transactionExpand,
        blobs: {
          select: {
            blobHash: true,
            ...blobExpand,
          },
          orderBy: {
            index: "asc",
          },
        },
      },
      orderBy: {
        index: "asc",
      },
      where: filters?.transactionFilters,
    },
  } satisfies Prisma.BlockSelect;
}

export const responseBlockSchema = baseBlockSchema.extend({
  transactions: z.array(
    baseTransactionSchema
      .omit({
        blockHash: true,
        blockNumber: true,
        blockTimestamp: true,
      })
      .partial()
      .required({ hash: true })
      .extend({
        blobs: z.array(
          baseBlobSchema.partial().required({ versionedHash: true })
        ),
      })
  ),
});

export type ResponseBlock = z.input<typeof responseBlockSchema>;

export function toResponseBlock(
  prismaBlock: CompletePrismaBlock
): ResponseBlock {
  const transactions = prismaBlock.transactions.map(
    ({ hash, blobs, ...prismaTx }) => ({
      hash,
      ...(isFullyDefined(prismaTx)
        ? {
            ...prismaTx,
            ...normalizePrismaTransactionFields(prismaTx),
            ...deriveTransactionFields({
              ...prismaTx,
              blobGasPrice: prismaBlock.blobGasPrice,
            }),
          }
        : {}),
      blobs: blobs.map(({ blobHash, blob }) =>
        blob
          ? normalizePrismaBlobFields({ versionedHash: blobHash, ...blob })
          : { versionedHash: blobHash }
      ),
    })
  );

  return {
    ...prismaBlock,
    transactions,
  };
}

export type BlockId = "hash" | "number" | "slot";
export type BlockIdField =
  | { type: "hash"; value: string }
  | { type: "number"; value: number }
  | { type: "slot"; value: number };

function buildBlockWhereClause(
  { type, value }: BlockIdField,
  filters: Filters
): Prisma.BlockWhereInput {
  switch (type) {
    case "hash": {
      return { hash: value };
    }
    case "number": {
      return { number: value, transactionForks: filters.blockType };
    }
    case "slot": {
      return { slot: value, transactionForks: filters.blockType };
    }
  }
}

export async function fetchBlock(
  blockId: BlockIdField,
  {
    blobStorageManager,
    prisma,
    filters,
    expands,
  }: {
    blobStorageManager: BlobStorageManager;
    prisma: BlobscanPrismaClient;
    filters: Filters;
    expands: Expands;
  }
): Promise<CompletePrismaBlock | undefined> {
  const select = createBlockSelect(expands);
  const where = buildBlockWhereClause(blockId, filters);

  const prismaBlock = (await prisma.block.findFirst({
    select,
    where,
  })) as unknown as CompletePrismaBlock | null;

  if (!prismaBlock) {
    return;
  }

  if (expands.blobData) {
    await Promise.all(
      prismaBlock.transactions
        .flatMap((tx) => tx.blobs)
        .map(async ({ blobHash, blob }) => {
          const dataStorageReferences = blob?.dataStorageReferences;
          if (dataStorageReferences) {
            const data = await retrieveBlobData(blobStorageManager, {
              dataStorageReferences,
              versionedHash: blobHash,
            });

            blob.data = data;
          }
        })
    );
  }

  return prismaBlock;
}
