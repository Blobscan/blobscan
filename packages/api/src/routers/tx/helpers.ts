import type { Prisma } from "@blobscan/db";
import { z } from "@blobscan/zod";

import type {
  ExpandedBlob,
  ExpandedBlock,
  Expands,
} from "../../middlewares/withExpands";
import type { Prettify } from "../../types";
import {
  deriveTransactionFields,
  normalizePrismaBlobFields,
  normalizePrismaTransactionFields,
} from "../../utils/transformers";
import {
  baseTransactionSchema,
  baseBlobSchema,
  baseBlockSchema,
} from "../../zod-schemas";

const transactionSelect = {
  hash: true,
  fromId: true,
  toId: true,
  blobGasUsed: true,
  blobAsCalldataGasUsed: true,
  gasPrice: true,
  maxFeePerBlobGas: true,
  blockHash: true,
  blockNumber: true,
  blockTimestamp: true,
  index: true,
  decodedFields: true,
  from: {
    select: {
      address: true,
      rollup: true,
    },
  },
} satisfies Prisma.TransactionSelect;

type PrismaTransaction = Prisma.TransactionGetPayload<{
  select: typeof transactionSelect;
}>;

export type CompletePrismaTransaction = Prettify<
  PrismaTransaction & {
    decodedFields: NonNullable<PrismaTransaction["decodedFields"]>;
    block: Prettify<
      Partial<ExpandedBlock> & {
        blobGasPrice: ExpandedBlock["blobGasPrice"];
      }
    >;
    blobs: Prettify<
      { blobHash: string } & {
        blob?: ExpandedBlob;
      }
    >[];
  }
>;
export function createTransactionSelect(expands: Expands) {
  const blockExpand = expands.block?.select;
  const blobExpand = expands.blob;

  return {
    ...transactionSelect,
    block: {
      select: {
        ...(blockExpand ?? {}),
        blobGasPrice: true,
      },
    },
    blobs: {
      select: {
        blobHash: true,
        ...(blobExpand ? { blob: blobExpand } : {}),
      },
      orderBy: {
        index: "asc",
      },
    },
  } satisfies Prisma.TransactionSelect;
}

export const responseTransactionSchema = baseTransactionSchema.extend({
  block: baseBlockSchema
    .omit({
      hash: true,
      number: true,
      timestamp: true,
    })
    .partial()
    .required({
      blobGasPrice: true,
    }),
  blobs: z.array(baseBlobSchema.partial().required({ versionedHash: true })),
});

export type ResponseTransaction = z.input<typeof responseTransactionSchema>;

export function toResponseTransaction(
  prismaTx: CompletePrismaTransaction
): ResponseTransaction {
  const { blobs: blobsOnTxs, block } = prismaTx;
  const normalizedFields = normalizePrismaTransactionFields(prismaTx);
  const derivedFields = deriveTransactionFields({
    ...prismaTx,
    blobGasPrice: block.blobGasPrice,
  });
  const blobs = blobsOnTxs.map(({ blobHash, blob }) =>
    blob
      ? normalizePrismaBlobFields({ versionedHash: blobHash, ...blob })
      : { versionedHash: blobHash }
  );

  return {
    ...prismaTx,
    ...normalizedFields,
    ...derivedFields,
    blobs,
  };
}
