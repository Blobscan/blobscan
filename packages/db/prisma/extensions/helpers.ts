import { Prisma } from "@prisma/client";

export type ZeroOpResult = { count: number }[];

export const helpersExtension = Prisma.defineExtension((prisma) =>
  prisma.$extends({
    name: "Helper Operations",
    model: {
      $allModels: {
        zero() {
          return prisma.$queryRaw<ZeroOpResult>`SELECT 0 as count`;
        },
      },
      block: {
        findLatest() {
          return prisma.block.findFirst({
            where: {
              transactionForks: {
                none: {},
              },
            },
            orderBy: { number: "desc" },
          });
        },
      },
    },
  })
);
