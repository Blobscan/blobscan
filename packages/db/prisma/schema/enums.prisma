enum BlobStorage {
  FILE_SYSTEM @map("file_system")
  GOOGLE      @map("google")
  POSTGRES    @map("postgres")
  SWARM       @map("swarm")
  WEAVEVM     @map("weavevm")
  S3          @map("s3")

  @@map("blob_storage")
}

enum Category {
  OTHER  @map("other")
  ROLLUP @map("rollup")

  @@map("category")
}

enum Rollup {
  ABSTRACT          @map("abstract")
  AEVO              @map("aevo")
  ANCIENT8          @map("ancient8")
  ARBITRUM          @map("arbitrum")
  ARENAZ            @map("arenaz")
  BASE              @map("base")
  BLAST             @map("blast")
  BOB               @map("bob")
  BOBA              @map("boba")
  CAMP              @map("camp")
  DEBANKCHAIN       @map("debankchain")
  ETHERNITY         @map("ethernity")
  FRAXTAL           @map("fraxtal")
  FUEL              @map("fuel")
  HASHKEY           @map("hashkey")
  HEMI              @map("hemi")
  HYPR              @map("hypr")
  INFINAEON         @map("infinaeon")
  INK               @map("ink")
  KARAK             @map("karak")
  KINTO             @map("kinto")
  KROMA             @map("kroma")
  LAMBDA            @map("lambda")
  LINEA             @map("linea")
  LISK              @map("lisk")
  MANTA             @map("manta")
  MANTLE            @map("mantle")
  METAL             @map("metal")
  METAMAIL          @map("metamail")
  METIS             @map("metis")
  MINT              @map("mint")
  MODE              @map("mode")
  MORPH             @map("morph")
  NAL               @map("nal")
  NANONNETWORK      @map("nanonnetwork")
  OPBNB             @map("opbnb")
  OPTIMISM          @map("optimism")
  OPTOPIA           @map("optopia")
  ORDERLY           @map("orderly")
  PANDASEA          @map("pandasea")
  PARADEX           @map("paradex")
  PARALLEL          @map("parallel")
  PHALA             @map("phala")
  PGN               @map("pgn")
  POLYNOMIAL        @map("polynomial")
  R0AR              @map("r0ar")
  RACE              @map("race")
  RARI              @map("rari")
  RIVER             @map("river")
  SCROLL            @map("scroll")
  SHAPE             @map("shape")
  SNAXCHAIN         @map("snaxchain")
  SONEIUM           @map("soneium")
  STARKNET          @map("starknet")
  SUPERLUMIO        @map("superlumio")
  SUPERSEED         @map("superseed")
  SWANCHAIN         @map("swanchain")
  SWELLCHAIN        @map("swellchain")
  TAIKO             @map("taiko")
  THEBINARYHOLDINGS @map("thebinaryholdings")
  UNICHAIN          @map("unichain")
  WORLD             @map("world")
  XGA               @map("xga")
  ZERONETWORK       @map("zeronetwork")
  ZIRCUIT           @map("zircuit")
  ZORA              @map("zora")
  ZKSYNC            @map("zksync")

  @@map("rollup")
}
