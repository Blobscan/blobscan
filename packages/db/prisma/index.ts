import { PrismaClient } from "@prisma/client";

import { baseExtension, statsExtension } from "./extensions";

const globalForPrisma = globalThis as unknown as { prisma: PrismaClient };

const prisma_ =
  globalForPrisma.prisma ||
  new PrismaClient({
    log:
      process.env.NODE_ENV === "development"
        ? ["query", "error", "warn"]
        : ["error"],
  });

export const prisma = prisma_.$extends(baseExtension).$extends(statsExtension);

export type BlobscanPrismaClient = typeof prisma;

export * from "./types";
export * from "./utils/dates";

if (process.env.NODE_ENV !== "production") globalForPrisma.prisma = prisma_;
