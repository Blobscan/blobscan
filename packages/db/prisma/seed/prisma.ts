import { PrismaClient } from "@prisma/client";

import { baseExtension, statsExtension } from "../extensions";

export function createLocalPrisma() {
  return new PrismaClient().$extends(baseExtension).$extends(statsExtension);
}
