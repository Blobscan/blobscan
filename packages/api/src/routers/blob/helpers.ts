import type {
  BlobscanPrismaClient,
  EthUsdPrice,
  ExtendedBlobDataStorageReferenceSelect,
  Prisma,
} from "@blobscan/db";
import { EthUsdPriceModel } from "@blobscan/db/prisma/zod";
import { z } from "@blobscan/zod";

import type {
  ExpandedBlob,
  ExpandedBlock,
  ExpandedTransaction,
  Expands,
} from "../../middlewares/withExpands";
import type { Prettify } from "../../types";
import { isFullyDefined } from "../../utils";
import {
  normalizePrismaBlobDataStorageReferencesFields,
  normalizePrismaTransactionFields,
} from "../../utils";
import {
  baseBlobSchema,
  baseBlockSchema,
  baseTransactionSchema,
  prismaBlobOnTransactionSchema,
} from "../../zod-schemas";

const blobDataStorageReferenceSelect = {
  blobStorage: true,
  url: true,
} satisfies ExtendedBlobDataStorageReferenceSelect;

export const extendedBlobDataStorageReferenceArgs: Prisma.Args<
  BlobscanPrismaClient["blobDataStorageReference"],
  "findFirstOrThrow"
> = {
  select: blobDataStorageReferenceSelect,
  orderBy: {
    blobStorage: "asc",
  },
};

export const baseBlobSelect = {
  commitment: true,
  proof: true,
  size: true,
  usageSize: true,
  versionedHash: true,
} satisfies Prisma.BlobSelect;

export const baseBlobOnTransactionSelect = {
  blobHash: true,
  blockHash: true,
  blockNumber: true,
  blockTimestamp: true,
  index: true,
  txHash: true,
  txIndex: true,
} satisfies Prisma.BlobsOnTransactionsSelect;

export type ExtendedBlobDataStorageReference = NonNullable<
  Prisma.Result<
    BlobscanPrismaClient["blobDataStorageReference"],
    { select: typeof blobDataStorageReferenceSelect },
    "findFirst"
  >
>;
type PrismaBlob = Prisma.BlobGetPayload<{
  select: typeof baseBlobSelect;
}> & {
  dataStorageReferences: ExtendedBlobDataStorageReference[];
};

type PrismaBlobOnTransaction = Prisma.BlobsOnTransactionsGetPayload<{
  select: typeof baseBlobOnTransactionSelect;
}>;

export type CompletedPrismaBlob = Prettify<
  PrismaBlob & {
    transactions: Prettify<
      PrismaBlobOnTransaction & {
        block?: ExpandedBlock;
        transaction?: ExpandedTransaction & {
          block: { blobGasPrice: ExpandedBlock["blobGasPrice"] };
        };
      }
    >[];
  }
>;

export type CompletedPrismaBlobOnTransaction = Prettify<
  PrismaBlobOnTransaction & {
    blob: {
      dataStorageReferences: ExtendedBlobDataStorageReference[];
    } & ExpandedBlob;
    block?: ExpandedBlock;
    transaction?: ExpandedTransaction & {
      block: { blobGasPrice: ExpandedBlock["blobGasPrice"] };
    };
  }
>;
export function createBlobSelect(expands: Expands) {
  const blockExpand = expands.block ? { block: expands.block } : {};
  const txExpand = expands.transaction
    ? {
        transaction: {
          select: {
            ...expands.transaction.select,
            block: {
              select: {
                blobGasPrice: true,
              },
            },
          },
        },
      }
    : {};

  return {
    ...baseBlobSelect,
    dataStorageReferences: extendedBlobDataStorageReferenceArgs,
    transactions: {
      select: {
        ...baseBlobOnTransactionSelect,
        ...blockExpand,
        ...txExpand,
      },
      orderBy: {
        blockNumber: "desc",
      },
    },
  } satisfies Prisma.BlobSelect;
}

export function createBlobsOnTransactionsSelect(expands: Expands) {
  const blockExpand = expands.block ? { block: expands.block } : {};
  const txExpand = expands.transaction
    ? {
        transaction: {
          select: {
            ...expands.transaction.select,
            block: {
              select: {
                blobGasPrice: true,
              },
            },
          },
        },
      }
    : {};

  return {
    ...baseBlobOnTransactionSelect,
    blob: {
      select: {
        ...baseBlobSelect,
        dataStorageReferences: extendedBlobDataStorageReferenceArgs,
      },
    },
    ...txExpand,
    ...blockExpand,
  } satisfies Prisma.BlobsOnTransactionsSelect;
}

const responseBlobOnTxSchema = prismaBlobOnTransactionSchema
  .omit({ blobHash: true })
  .required({
    blockHash: true,
    blockNumber: true,
    blockTimestamp: true,
    index: true,
    txHash: true,
    txIndex: true,
  });

export const responseBlobSchema = baseBlobSchema.extend({
  transactions: z.array(
    responseBlobOnTxSchema
      .merge(
        baseTransactionSchema
          .omit({
            hash: true,
            blockNumber: true,
            blockTimestamp: true,
            blockHash: true,
          })
          .partial()
      )
      .extend({
        block: baseBlockSchema
          .omit({
            hash: true,
            number: true,
            timestamp: true,
          })
          .partial({
            blobGasBaseUsdFee: true,
            blobGasUsdPrice: true,
          })
          .optional(),
        ethUsdPrice: EthUsdPriceModel.shape.price.optional(),
      })
  ),
});

export const responseBlobOnTransactionSchema = responseBlobOnTxSchema
  .merge(baseBlobSchema)
  .extend({
    block: baseBlockSchema
      .omit({
        hash: true,
        number: true,
        timestamp: true,
      })
      .optional(),
    transaction: baseTransactionSchema
      .omit({
        hash: true,
        blockNumber: true,
        blockTimestamp: true,
        blockHash: true,
      })
      .optional(),
  });

export type ResponseBlob = z.input<typeof responseBlobSchema>;
export type ResponseBlobOnTransaction = z.input<
  typeof responseBlobOnTransactionSchema
>;

export function toResponseBlob(
  {
    dataStorageReferences,
    transactions: blobOnTxs,
    ...prismaBlob
  }: CompletedPrismaBlob,
  ethUsdPrices: EthUsdPrice["price"][]
): ResponseBlob {
  const responseBlob: ResponseBlob = {
    ...prismaBlob,
    dataStorageReferences: normalizePrismaBlobDataStorageReferencesFields(
      dataStorageReferences
    ),
    transactions: blobOnTxs.map(
      (
        {
          block: { computeUsdFields, ...restBlock } = {},
          transaction: prismaTx = {},
          ...restBlobOnTx
        },
        i
      ) => {
        const txEthUsdPrice = ethUsdPrices[i];

        const blockUsdFields =
          txEthUsdPrice && computeUsdFields
            ? computeUsdFields(txEthUsdPrice)
            : undefined;
        const txBlock = {
          ...restBlock,
          ...(blockUsdFields ?? {}),
        };

        let responseBlobOnTx: ResponseBlob["transactions"][number] = {
          ...restBlobOnTx,
          ...txBlock,
          ethUsdPrice: txEthUsdPrice?.toNumber(),
        };

        if (isFullyDefined(prismaTx)) {
          const {
            block,
            decodedFields,
            from,
            fromId,
            toId,
            computeBlobGasBaseFee,
            computeUsdFields,
            ...restPrismaTx
          } = prismaTx;

          responseBlobOnTx = {
            ...responseBlobOnTx,
            ...restPrismaTx,
            ...normalizePrismaTransactionFields({
              decodedFields,
              from,
              fromId,
              toId,
            }),
            blobGasBaseFee: computeBlobGasBaseFee(block.blobGasPrice),
            ...(txEthUsdPrice
              ? computeUsdFields({
                  blobGasPrice: block.blobGasPrice,
                  ethUsdPrice: txEthUsdPrice,
                })
              : {}),
          };
        }

        return responseBlobOnTx;
      }
    ),
  };

  return responseBlob;
}

export function toResponseBlobOnTransaction(
  prismaBlobOnTransaction: CompletedPrismaBlobOnTransaction,
  ethUsdPrice?: EthUsdPrice["price"]
): ResponseBlobOnTransaction {
  const {
    blobHash,
    blob: { dataStorageReferences, ...blob },
    transaction,
    block,
    ...restBlobOnTransaction
  } = prismaBlobOnTransaction;

  const responseBlobOnTx: ResponseBlobOnTransaction = {
    ...blob,
    versionedHash: blobHash,
    dataStorageReferences: normalizePrismaBlobDataStorageReferencesFields(
      dataStorageReferences
    ),
    ...restBlobOnTransaction,
  };

  if (transaction) {
    const {
      block: txBlock,
      decodedFields,
      from,
      fromId,
      toId,
      computeBlobGasBaseFee,
      computeUsdFields,
      ...restTx
    } = transaction;

    responseBlobOnTx.transaction = {
      ...restTx,
      blobGasPrice: txBlock.blobGasPrice,
      ...normalizePrismaTransactionFields({
        decodedFields,
        from,
        fromId,
        toId,
      }),
      blobGasBaseFee: computeBlobGasBaseFee(txBlock.blobGasPrice),
      ...(ethUsdPrice
        ? computeUsdFields({
            blobGasPrice: txBlock.blobGasPrice,
            ethUsdPrice,
          })
        : {}),
    };
  }

  if (block) {
    const { computeUsdFields, ...restBlock } = block;
    responseBlobOnTx.block = {
      ...restBlock,
      ...(ethUsdPrice ? computeUsdFields(ethUsdPrice) : {}),
    };
  }

  return responseBlobOnTx;
}
