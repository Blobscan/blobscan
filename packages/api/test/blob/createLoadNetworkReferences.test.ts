import { TRPCError } from "@trpc/server";
import { beforeAll, describe, expect, it } from "vitest";

import { BlobStorage } from "@blobscan/db/prisma/enums";
import { testValidError } from "@blobscan/test";

import type { TRPCContext } from "../../src";
import { createTestContext, unauthorizedRPCCallTest } from "../helpers";
import { createBlobCaller } from "./caller";
import type { BlobCaller } from "./caller";

describe("createLoadNetworkReferences", () => {
  let authorizedBlobCaller: BlobCaller;
  let unauthorizedBlobCaller: BlobCaller;
  let unauthorizedCtx: TRPCContext;
  let authorizedContext: TRPCContext;

  const blobHashes = ["blobHash001", "blobHash002", "blobHash003"];
  const where = {
    AND: [
      {
        blobHash: {
          in: blobHashes,
        },
      },
      {
        blobStorage: BlobStorage.WEAVEVM,
      },
    ],
  };

  beforeAll(async () => {
    authorizedContext = await createTestContext({
      apiClient: "load-network",
    });
    unauthorizedCtx = await createTestContext();

    authorizedBlobCaller = createBlobCaller(authorizedContext);
    unauthorizedBlobCaller = createBlobCaller(unauthorizedCtx);
  });

  describe("when authorized", () => {
    it("should insert references correctly", async () => {
      const blobReferencesBefore =
        await unauthorizedCtx.prisma.blobDataStorageReference.findMany({
          where,
        });

      expect(
        blobReferencesBefore,
        "There should be no blob weavevm references initially"
      ).toEqual([]);

      await authorizedBlobCaller.createWeaveVMReferences({
        blobHashes,
      });

      const blobReferencesAfter =
        await unauthorizedCtx.prisma.blobDataStorageReference
          .findMany({
            where,
          })
          .then((refs) =>
            refs
              .map(({ blobHash }) => blobHash)
              .sort((a, b) => a.localeCompare(b))
          );

      expect(
        blobReferencesAfter,
        "References should have been inserted"
      ).toEqual(blobHashes.sort((a, b) => a.localeCompare(b)));
    });

    it("should skip already existing references correctly", async () => {
      await unauthorizedCtx.prisma.blobDataStorageReference.createMany({
        data: blobHashes.map((blobHash) => ({
          blobHash,
          blobStorage: BlobStorage.WEAVEVM,
          dataReference: blobHash,
        })),
      });

      await authorizedBlobCaller.createWeaveVMReferences({
        blobHashes,
      });

      const blobReferencesAfter =
        await unauthorizedCtx.prisma.blobDataStorageReference
          .findMany({
            where,
          })
          .then((refs) =>
            refs
              .map(({ blobHash }) => blobHash)
              .sort((a, b) => a.localeCompare(b))
          );

      expect(blobReferencesAfter).toEqual(
        blobHashes.sort((a, b) => a.localeCompare(b))
      );
    });

    it("should be called with an empty blob hashes array correctly", async () => {
      await expect(
        authorizedBlobCaller.createWeaveVMReferences({
          blobHashes: [],
        })
      ).resolves.toBeUndefined();
    });

    testValidError(
      "should fail when one or more provided blobs do not exist",
      async () => {
        await authorizedBlobCaller.createWeaveVMReferences({
          blobHashes: ["nonExistingBlobHash"],
        });
      },
      TRPCError,
      {
        checkCause: true,
      }
    );

    unauthorizedRPCCallTest(() =>
      unauthorizedBlobCaller.createWeaveVMReferences({
        blobHashes,
      })
    );
  });
});
