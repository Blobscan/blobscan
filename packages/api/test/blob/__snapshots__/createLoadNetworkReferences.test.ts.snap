// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createLoadNetworkReferences > when authorized > should fail when one or more provided blobs do not exist 1`] = `"Couldn't found the following blobs: \\"nonExistingBlobHash\\""`;

exports[`createLoadNetworkReferences > when authorized > should fail when one or more provided blobs do not exist 2`] = `undefined`;
